diff --git a/text-icu.cabal b/text-icu.cabal
index 27a5b13..24a63d1 100644
--- a/text-icu.cabal
+++ b/text-icu.cabal
@@ -68,6 +68,11 @@ tested-with:
   GHC == 8.0.2
   GHC == 7.10.3
 
+flag homebrew
+  Description: Assume homebrew on macOS. Automatically add /usr/local/opt/ and /usr/homebrew/opt/ paths to extra-lib-dirs and include-dirs.
+  Default:     True
+  Manual:      True
+
 library
   default-language:  Haskell98
   build-depends:
@@ -122,7 +127,7 @@ library
   c-sources: cbits/text_icu.c
   cc-options: -Wall -Wextra -pedantic -Wno-deprecated
   include-dirs: include
-  if os(darwin)
+  if os(darwin) && flag(homebrew)
     extra-lib-dirs:
       /usr/local/opt/icu4c/lib
       /opt/homebrew/opt/icu4c/lib
