diff --git a/src/Network/Bugsnag/BugsnagRequestHeaders.hs b/src/Network/Bugsnag/BugsnagRequestHeaders.hs
index 161c99b..d3c674f 100644
--- a/src/Network/Bugsnag/BugsnagRequestHeaders.hs
+++ b/src/Network/Bugsnag/BugsnagRequestHeaders.hs
@@ -24,7 +24,7 @@ instance ToJSON BugsnagRequestHeaders where
     toJSON = object . map headerToKeyValue . unBugsnagRequestHeaders
     toEncoding = pairs . foldMap headerToKeyValue . unBugsnagRequestHeaders
 
-headerToKeyValue :: KeyValue kv => (CI ByteString, ByteString) -> kv
+headerToKeyValue :: KeyValue e kv => (CI ByteString, ByteString) -> kv
 headerToKeyValue (name, value) =
     fromText (TE.decodeUtf8 (CI.original name)) .= String (TE.decodeUtf8 value)
 
